imod/protocols/protocol_ctfEstimation_automatic.py

```diff
@@ -386,56 +393,60 @@ def ctfEstimation(self, tsId, expDefoci):
             self.runProgram('ctfplotter', paramsCtfPlotter)
 
         except Exception as e:
-            self._failedTs.append(tsId)
+            self._failedItems.append(tsId)
             self.error(f'ctfplotter execution failed for tsId {tsId} -> {e}')
 
     def createOutputStep(self, tsId, outputSetName=OUTPUT_CTF_SERIE):
         ts = self.tsDict[tsId]
-        if tsId in self._failedTs:
-            self.createOutputFailedSet(ts)
-        else:
-            defocusFilePath = self.getExtraOutFile(tsId, ext=DEFOCUS_EXT)
-            if os.path.exists(defocusFilePath):
-                output = self.getOutputSetOfCTFTomoSeries(outputSetName)
-                defocusFileFlag = utils.getDefocusFileFlag(defocusFilePath)
-
-                newCTFTomoSeries = CTFTomoSeries(tsId=tsId)
-                newCTFTomoSeries.copyInfo(ts)
-                newCTFTomoSeries.setTiltSeries(ts)
-                newCTFTomoSeries.setIMODDefocusFileFlag(defocusFileFlag)
-                newCTFTomoSeries.setNumberOfEstimationsInRange(None)
-                output.append(newCTFTomoSeries)
-
-                self.parseTSDefocusFile(ts, defocusFilePath, newCTFTomoSeries)
 
-                if not (newCTFTomoSeries.getIsDefocusUDeviationInRange() and
-                        newCTFTomoSeries.getIsDefocusVDeviationInRange()):
-                    newCTFTomoSeries.setEnabled(False)
-
-                output.update(newCTFTomoSeries)
-                output.write()
-                self._store(output)
+        with self._lock:
+            if tsId in self._failedItems:
+                self.createOutputFailedSet(ts)
+            else:
+                defocusFilePath = self.getExtraOutFile(tsId, ext=DEFOCUS_EXT)
+                if os.path.exists(defocusFilePath):
+                    output = self.getOutputSetOfCTFTomoSeries(self.getInputSet(pointer=True),
+                                                              outputSetName)
+                    newCTFTomoSeries = CTFTomoSeries(tsId=tsId)
+                    newCTFTomoSeries.copyInfo(ts)
+                    newCTFTomoSeries.setTiltSeries(ts)
+
+                    # flags below will be updated in parseTSDefocusFile
+                    newCTFTomoSeries.setIMODDefocusFileFlag(1)
+                    newCTFTomoSeries.setNumberOfEstimationsInRange(0)
+                    output.append(newCTFTomoSeries)
+
+                    self.parseTSDefocusFile(ts, defocusFilePath, newCTFTomoSeries)
+
+                    output.update(newCTFTomoSeries)
+                    output.write()
+                    self._store(output)
```

Comment:
I can see you put back these two:
output.write()
self._store(output)
Is there a reason? Protocol works without these redundant calls. Write and store are called in the closeOutputSet step for all outputs

Reply:
Hi, @azazellochg. Yes, the reason is (what we call) the semi-streamified version of the protocol, which means that the outputs are generated during execution time instead of at the end of the protocol execution, even if they are not streamified protocols yet. Without that calls or not all of them the outputs may be not generated or generated empty, so they can't be visualized and so on until they are filled at the end of the protocol execution.

Reply:
On the other hand, the update call is necessary to update some critical attributes of the set, such as the size in case of size updating (e. g., a binning operation) or if the set if still heterogeneous, something that may change with, for example, a views exclusion operation.